components/clp-py-utils/clp_py_utils/s3_utils.py

```diff
@@ -1,18 +1,208 @@
+) -> Tuple[bool, List[str]]:
+    """
+    Generates Docker container authentication options for AWS S3 access based on the given type.
+    Handles authentication methods that require extra configuration (profile, env_vars).
+
+    :param clp_config: CLPConfig containing storage configurations.
+    :param type: Type of calling container (compression, log_viewer, or query).
+    :return: Tuple of (whether aws config mount is needed, credential env_vars to set).
+    :raises: ValueError if environment variables are not set correctly.
+    """
+    if component_type not in [
+        COMPRESSION_SCHEDULER_COMPONENT_NAME,
+        COMPRESSION_WORKER_COMPONENT_NAME,
+        LOG_VIEWER_WEBUI_COMPONENT_NAME,
+        QUERY_SCHEDULER_COMPONENT_NAME,
+        QUERY_WORKER_COMPONENT_NAME,
+    ]:
+        raise ValueError(f"Component type {component_type} is not valid.")
+
+    storages_by_component_type = []
+    if (
+        COMPRESSION_SCHEDULER_COMPONENT_NAME == component_type
+        or COMPRESSION_WORKER_COMPONENT_NAME == component_type
+    ):
+        storages_by_component_type = [clp_config.logs_input, clp_config.archive_output.storage]
+    elif LOG_VIEWER_WEBUI_COMPONENT_NAME == component_type:
+        storages_by_component_type = [clp_config.stream_output.storage]
+    elif (
+        QUERY_SCHEDULER_COMPONENT_NAME == component_type
+        or QUERY_WORKER_COMPONENT_NAME == component_type
+    ):
+        storages_by_component_type = [
+            clp_config.logs_input,
+            clp_config.archive_output.storage,
+            clp_config.stream_output.storage,
+        ]
+
+    config_mount = False
+    add_env_vars = False
+
+    for storage in storages_by_component_type:
+        if StorageType.S3 == storage.type:
+            auth = storage.s3_config.aws_authentication
+            if AwsAuthType.profile == auth.type:
+                config_mount = True
+            elif AwsAuthType.env_vars == auth.type:
+                add_env_vars = True
+
+    credentials_env_vars = []
+
+    if add_env_vars:
+        access_key_id = os.getenv("AWS_ACCESS_KEY_ID")
+        secret_access_key = os.getenv("AWS_SECRET_ACCESS_KEY")
+        if access_key_id and secret_access_key:
+            credentials_env_vars.extend(
+                (
+                    f"AWS_ACCESS_KEY_ID={access_key_id}",
+                    f"AWS_SECRET_ACCESS_KEY={secret_access_key}",
+                )
+            )
+        else:
+            raise ValueError(
+                "AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables not set"
+            )
+        if os.getenv("AWS_SESSION_TOKEN"):
+            raise ValueError(
+                "AWS_SESSION_TOKEN not supported for environmental variable credentials."
+            )
+
+    return (config_mount, credentials_env_vars)
+
+
+def _create_s3_client(s3_config: S3Config) -> boto3.client:
+    config = Config(retries=dict(total_max_attempts=3, mode="adaptive"))
```

Comment:
let's make config an optional argument (with None as default), and let caller passes it in.
For now, I don't see issue applying `total_max_attempts=3` to ever client, but it's possible there will be some configs we want to apply case-by-case instead of to all clients